File: genai_telemetry/exporters/prometheus.py
"""
Prometheus Push Gateway exporter.
"""

import base64
import logging
import threading
import urllib.request
from typing import Any, Dict

from genai_telemetry.exporters.base import BaseExporter

logger = logging.getLogger("genai_telemetry.exporters.prometheus")


class PrometheusExporter(BaseExporter):
    """Sends metrics to Prometheus Push Gateway."""
    
    def __init__(
        self,
        pushgateway_url: str = "http://localhost:9091",
        job_name: str = "genai_telemetry",
        username: str = None,
        password: str = None
    ):
        """
        Initialize Prometheus exporter.
        
        Args:
            pushgateway_url: URL of the Prometheus Push Gateway
            job_name: Job name for metrics grouping
            username: Username for basic auth (optional)
            password: Password for basic auth (optional)
        """
        self.pushgateway_url = pushgateway_url.rstrip("/")
        self.job_name = job_name
        self.username = username
        self.password = password
        
        # Metrics storage
        self._metrics: Dict[str, Dict] = {
            "llm_requests_total": {"type": "counter", "value": 0},
            "llm_duration_seconds": {"type": "histogram", "values": []},
            "llm_tokens_total": {"type": "counter", "input": 0, "output": 0},
            "llm_errors_total": {"type": "counter", "value": 0},
        }
        self._lock = threading.Lock()
    
    def _get_headers(self) -> dict:
        """Build authentication headers."""
        headers = {"Content-Type": "text/plain"}
        if self.username and self.password:
            credentials = base64.b64encode(f"{self.username}:{self.password}".encode()).decode()
            headers["Authorization"] = f"Basic {credentials}"
        return headers
    
    def export(self, span_data: Dict[str, Any]) -> bool:
        """Update metrics and push to gateway."""
        with self._lock:
            # Update counters
            self._metrics["llm_requests_total"]["value"] += 1
            
            if span_data.get("is_error"):
                self._metrics["llm_errors_total"]["value"] += 1
            
            # Update tokens
            self._metrics["llm_tokens_total"]["input"] += span_data.get("input_tokens", 0)
            self._metrics["llm_tokens_total"]["output"] += span_data.get("output_tokens", 0)
            
            # Update duration histogram
            duration_sec = span_data.get("duration_ms", 0) / 1000
            self._metrics["llm_duration_seconds"]["values"].append(duration_sec)
        
        return self._push_metrics(span_data)
    
    def _push_metrics(self, span_data: Dict[str, Any]) -> bool:
        """Push metrics to Prometheus Push Gateway."""
        model = span_data.get("model_name", "unknown")
        provider = span_data.get("model_provider", "unknown")
        workflow = span_data.get("workflow_name", "unknown")
        
        # Build Prometheus text format
        lines = []
        
        with self._lock:
            # Counter: requests
            lines.append('# TYPE llm_requests_total counter')
            lines.append(f'llm_requests_total{{model="{model}",provider="{provider}",workflow="{workflow}"}} {self._metrics["llm_requests_total"]["value"]}')
            
            # Counter: errors
            lines.append('# TYPE llm_errors_total counter')
            lines.append(f'llm_errors_total{{model="{model}",provider="{provider}",workflow="{workflow}"}} {self._metrics["llm_errors_total"]["value"]}')
            
            # Counter: tokens
            lines.append('# TYPE llm_input_tokens_total counter')
            lines.append(f'llm_input_tokens_total{{model="{model}",provider="{provider}"}} {self._metrics["llm_tokens_total"]["input"]}')
            lines.append('# TYPE llm_output_tokens_total counter')
            lines.append(f'llm_output_tokens_total{{model="{model}",provider="{provider}"}} {self._metrics["llm_tokens_total"]["output"]}')
            
            # Gauge: last duration
            duration_sec = span_data.get("duration_ms", 0) / 1000
            lines.append('# TYPE llm_duration_seconds gauge')
            lines.append(f'llm_duration_seconds{{model="{model}",provider="{provider}"}} {duration_sec}')
        
        payload = "\n".join(lines) + "\n"
        
        url = f"{self.pushgateway_url}/metrics/job/{self.job_name}"
        data = payload.encode("utf-8")
        
        req = urllib.request.Request(
            url,
            data=data,
            headers=self._get_headers(),
            method="POST"
        )
        
        try:
            with urllib.request.urlopen(req, timeout=10) as resp:
                return resp.status in (200, 202)
        except Exception as e:
            logger.error(f"Prometheus Push Error: {e}")
            return False

File: js/src/exporters/prometheus.ts
/**
 * Prometheus Push Gateway exporter.
 */

import type { SpanData } from '../types';
import { BaseExporter } from './base';

export interface PrometheusExporterOptions {
  pushgatewayUrl?: string;
  jobName?: string;
}

export class PrometheusExporter extends BaseExporter {
  private pushgatewayUrl: string;
  private jobName: string;

  // Metrics storage
  private metrics: {
    llmDuration: Map<string, number[]>;
    llmTokens: Map<string, { input: number; output: number }>;
    llmErrors: Map<string, number>;
    embeddingDuration: Map<string, number[]>;
    retrieverDuration: Map<string, number[]>;
    retrieverDocs: Map<string, number[]>;
  };

  constructor(options: PrometheusExporterOptions = {}) {
    super();
    this.pushgatewayUrl = (options.pushgatewayUrl || 'http://localhost:9091').replace(/\/$/, '');
    this.jobName = options.jobName || 'genai_telemetry';

    this.metrics = {
      llmDuration: new Map(),
      llmTokens: new Map(),
      llmErrors: new Map(),
      embeddingDuration: new Map(),
      retrieverDuration: new Map(),
      retrieverDocs: new Map(),
    };
  }

  private updateMetrics(span: SpanData): void {
    const key = `${span.workflow_name || 'default'}:${span.model_name || span.name}`;

    switch (span.span_type) {
      case 'LLM': {
        if (!this.metrics.llmDuration.has(key)) {
          this.metrics.llmDuration.set(key, []);
        }
        this.metrics.llmDuration.get(key)!.push(span.duration_ms);

        const tokens = this.metrics.llmTokens.get(key) || { input: 0, output: 0 };
        tokens.input += span.input_tokens || 0;
        tokens.output += span.output_tokens || 0;
        this.metrics.llmTokens.set(key, tokens);

        if (span.is_error) {
          this.metrics.llmErrors.set(key, (this.metrics.llmErrors.get(key) || 0) + 1);
        }
        break;
      }
      case 'EMBEDDING': {
        if (!this.metrics.embeddingDuration.has(key)) {
          this.metrics.embeddingDuration.set(key, []);
        }
        this.metrics.embeddingDuration.get(key)!.push(span.duration_ms);
        break;
      }
      case 'RETRIEVER': {
        const rKey = `${span.workflow_name || 'default'}:${span.vector_store || span.name}`;
        if (!this.metrics.retrieverDuration.has(rKey)) {
          this.metrics.retrieverDuration.set(rKey, []);
        }
        this.metrics.retrieverDuration.get(rKey)!.push(span.duration_ms);

        if (!this.metrics.retrieverDocs.has(rKey)) {
          this.metrics.retrieverDocs.set(rKey, []);
        }
        this.metrics.retrieverDocs.get(rKey)!.push(span.documents_retrieved || 0);
        break;
      }
    }
  }

  private buildMetricsPayload(): string {
    const lines: string[] = [];

    // LLM duration histogram (simplified as gauge with count and sum)
    lines.push('# HELP genai_llm_duration_seconds LLM call duration');
    lines.push('# TYPE genai_llm_duration_seconds summary');
    for (const [key, durations] of this.metrics.llmDuration) {
      const [workflow, model] = key.split(':');
      const labels = `workflow="${workflow}",model="${model}"`;
      const sum = durations.reduce((a, b) => a + b, 0) / 1000;
      lines.push(`genai_llm_duration_seconds_sum{${labels}} ${sum}`);
      lines.push(`genai_llm_duration_seconds_count{${labels}} ${durations.length}`);
    }

    // LLM tokens
    lines.push('# HELP genai_llm_tokens_total Total tokens used');
    lines.push('# TYPE genai_llm_tokens_total counter');
    for (const [key, tokens] of this.metrics.llmTokens) {
      const [workflow, model] = key.split(':');
      const labels = `workflow="${workflow}",model="${model}"`;
      lines.push(`genai_llm_tokens_total{${labels},type="input"} ${tokens.input}`);
      lines.push(`genai_llm_tokens_total{${labels},type="output"} ${tokens.output}`);
    }

    // LLM errors
    lines.push('# HELP genai_llm_errors_total Total LLM errors');
    lines.push('# TYPE genai_llm_errors_total counter');
    for (const [key, count] of this.metrics.llmErrors) {
      const [workflow, model] = key.split(':');
      const labels = `workflow="${workflow}",model="${model}"`;
      lines.push(`genai_llm_errors_total{${labels}} ${count}`);
    }

    // Embedding duration
    lines.push('# HELP genai_embedding_duration_seconds Embedding call duration');
    lines.push('# TYPE genai_embedding_duration_seconds summary');
    for (const [key, durations] of this.metrics.embeddingDuration) {
      const [workflow, model] = key.split(':');
      const labels = `workflow="${workflow}",model="${model}"`;
      const sum = durations.reduce((a, b) => a + b, 0) / 1000;
      lines.push(`genai_embedding_duration_seconds_sum{${labels}} ${sum}`);
      lines.push(`genai_embedding_duration_seconds_count{${labels}} ${durations.length}`);
    }

    // Retriever duration
    lines.push('# HELP genai_retriever_duration_seconds Retriever call duration');
    lines.push('# TYPE genai_retriever_duration_seconds summary');
    for (const [key, durations] of this.metrics.retrieverDuration) {
      const [workflow, store] = key.split(':');
      const labels = `workflow="${workflow}",vector_store="${store}"`;
      const sum = durations.reduce((a, b) => a + b, 0) / 1000;
      lines.push(`genai_retriever_duration_seconds_sum{${labels}} ${sum}`);
      lines.push(`genai_retriever_duration_seconds_count{${labels}} ${durations.length}`);
    }

    // Retriever documents
    lines.push('# HELP genai_retriever_documents_total Documents retrieved');
    lines.push('# TYPE genai_retriever_documents_total counter');
    for (const [key, docs] of this.metrics.retrieverDocs) {
      const [workflow, store] = key.split(':');
      const labels = `workflow="${workflow}",vector_store="${store}"`;
      const total = docs.reduce((a, b) => a + b, 0);
      lines.push(`genai_retriever_documents_total{${labels}} ${total}`);
    }

    return lines.join('\n') + '\n';
  }

  async export(spanData: SpanData): Promise<boolean> {
    try {
      this.updateMetrics(spanData);
      const payload = this.buildMetricsPayload();

      const response = await fetch(
        `${this.pushgatewayUrl}/metrics/job/${this.jobName}`,
        {
          method: 'PUT',
          headers: { 'Content-Type': 'text/plain' },
          body: payload,
        }
      );

      return response.ok;
    } catch (error) {
      console.error('Prometheus Push Gateway Error:', error);
      return false;
    }
  }

  async healthCheck(): Promise<boolean> {
    try {
      const response = await fetch(`${this.pushgatewayUrl}/-/healthy`);
      return response.ok;
    } catch {
      return false;
    }
  }
}
